manifests/init.pp

```diff
@@ -460,17 +462,32 @@
       }
     }
 
-    python::pip { 'patroni':
-      ensure      => $version,
-      virtualenv  => $install_dir,
-      environment => ["PIP_PREFIX=${install_dir}"],
-      before      => File['patroni_config'],
-    }
+    if $custom_pip_provider != undef {
+      python::pip { 'patroni':
```

Comment:
One way to simplify this code and reduce duplication is just have this:

```
python::pip { 'patroni':
  ...
  pip_provider => $custom_pip_provider,
}
```
Instead of having the resources defined in 2 places it's better the have the parameters passed to a single resource based on conditions, reduces duplication.  If the goal with custom_pip_provider is to not define virtual value then you could do something like this:

```
if $custom_pip_provider {
  $virtualenv = undef
} else {
  $virtualenv = $install_dir
}
```

Then do something like :

```
python::pip { 'patroni':
...
  pip_provider => $custom_pip_provider,
  virtualenv => $virtualenv,
..
}
```

And something similar for dependency_params

Reply:
Totally agree - for some reason I was thinking that specifying the var as 'undef' would cause the default to not be used but that's the very point. I will update.